File: agent/src/InfraControl.Agent.Core/Services/BackoffPolicy.cs
namespace InfraControl.Agent.Core.Services;

public sealed class BackoffPolicy
{
    public TimeSpan GetDelay(int consecutiveFailures)
    {
        if (consecutiveFailures <= 0)
        {
            return TimeSpan.Zero;
        }

        int multiplier = 1 << Math.Min(consecutiveFailures - 1, 5);
        int seconds = Math.Min(120, 5 * multiplier);

        return TimeSpan.FromSeconds(seconds);
    }
}

File: agent/src/services/backoff-policy.mjs
export function createBackoffPolicy() {
  return {
    getDelaySeconds(consecutiveFailures) {
      if (consecutiveFailures <= 0) {
        return 0;
      }

      const multiplier = 1 << Math.min(consecutiveFailures - 1, 5);
      return Math.min(120, 5 * multiplier);
    },
  };
}
